File: android/app/src/main/java/com/dailymemory/data/service/ExportService.kt
package com.dailymemory.data.service

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.content.FileProvider
import com.dailymemory.domain.model.Memory
import com.dailymemory.domain.model.Person
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileWriter
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ExportService @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val exportDir: File
        get() = File(context.cacheDir, "exports").apply { mkdirs() }

    private val dateFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME

    /**
     * Export memories to JSON format
     */
    suspend fun exportToJson(
        memories: List<Memory>,
        persons: List<Person> = emptyList()
    ): Result<ExportResult> = withContext(Dispatchers.IO) {
        try {
            val timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"))
            val fileName = "dailymemory_export_$timestamp.json"
            val file = File(exportDir, fileName)

            val jsonObject = JSONObject().apply {
                put("exportDate", LocalDateTime.now().format(dateFormatter))
                put("version", "1.0")
                put("memoriesCount", memories.size)
                put("personsCount", persons.size)

                // Memories array
                put("memories", JSONArray().apply {
                    memories.forEach { memory ->
                        put(memoryToJson(memory))
                    }
                })

                // Persons array
                put("persons", JSONArray().apply {
                    persons.forEach { person ->
                        put(personToJson(person))
                    }
                })
            }

            FileWriter(file).use { writer ->
                writer.write(jsonObject.toString(2))
            }

            val uri = FileProvider.getUriForFile(
                context,
                "${context.packageName}.fileprovider",
                file
            )

            Result.success(ExportResult(
                uri = uri,
                fileName = fileName,
                mimeType = "application/json",
                itemCount = memories.size
            ))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Export memories to CSV format
     */
    suspend fun exportToCsv(memories: List<Memory>): Result<ExportResult> = withContext(Dispatchers.IO) {
        try {
            val timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"))
            val fileName = "dailymemory_export_$timestamp.csv"
            val file = File(exportDir, fileName)

            FileWriter(file).use { writer ->
                // Write header
                writer.write("ID,Date,Time,Content,Category,Importance,Location,Amount,Persons,Tags,IsLocked\n")

                // Write data rows
                memories.forEach { memory ->
                    val row = buildString {
                        append(escapeCsv(memory.id))
                        append(",")
                        append(memory.recordedAt.format(DateTimeFormatter.ISO_LOCAL_DATE))
                        append(",")
                        append(memory.recordedAt.format(DateTimeFormatter.ofPattern("HH:mm:ss")))
                        append(",")
                        append(escapeCsv(memory.content))
                        append(",")
                        append(memory.category.name)
                        append(",")
                        append(memory.importance)
                        append(",")
                        append(escapeCsv(memory.extractedLocation ?: ""))
                        append(",")
                        append(memory.extractedAmount?.toString() ?: "")
                        append(",")
                        append(escapeCsv(memory.extractedPersons.joinToString("; ")))
                        append(",")
                        append(escapeCsv(memory.extractedTags.joinToString("; ")))
                        append(",")
                        append(memory.isLocked)
                        append("\n")
                    }
                    writer.write(row)
                }
            }

            val uri = FileProvider.getUriForFile(
                context,
                "${context.packageName}.fileprovider",
                file
            )

            Result.success(ExportResult(
                uri = uri,
                fileName = fileName,
                mimeType = "text/csv",
                itemCount = memories.size
            ))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Export memories for a specific date range
     */
    suspend fun exportDateRange(
        memories: List<Memory>,
        startDate: LocalDateTime,
        endDate: LocalDateTime,
        format: ExportFormat
    ): Result<ExportResult> {
        val filteredMemories = memories.filter { memory ->
            memory.recordedAt >= startDate && memory.recordedAt <= endDate
        }

        return when (format) {
            ExportFormat.JSON -> exportToJson(filteredMemories)
            ExportFormat.CSV -> exportToCsv(filteredMemories)
        }
    }

    /**
     * Create share intent for export file
     */
    fun createShareIntent(exportResult: ExportResult): Intent {
        return Intent(Intent.ACTION_SEND).apply {
            type = exportResult.mimeType
            putExtra(Intent.EXTRA_STREAM, exportResult.uri)
            putExtra(Intent.EXTRA_SUBJECT, "DailyMemory Export - ${exportResult.fileName}")
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
    }

    /**
     * Clean up old export files
     */
    fun cleanupExports() {
        exportDir.listFiles()?.forEach { file ->
            // Delete files older than 1 day
            if (System.currentTimeMillis() - file.lastModified() > 24 * 60 * 60 * 1000) {
                file.delete()
            }
        }
    }

    private fun memoryToJson(memory: Memory): JSONObject {
        return JSONObject().apply {
            put("id", memory.id)
            put("content", memory.content)
            put("category", memory.category.name)
            put("importance", memory.importance)
            put("isLocked", memory.isLocked)
            put("recordedAt", memory.recordedAt.format(dateFormatter))
            put("createdAt", memory.createdAt.format(dateFormatter))
            put("updatedAt", memory.updatedAt.format(dateFormatter))

            memory.extractedLocation?.let { put("extractedLocation", it) }
            memory.extractedAmount?.let { put("extractedAmount", it) }
            memory.recordedLatitude?.let { put("latitude", it) }
            memory.recordedLongitude?.let { put("longitude", it) }

            if (memory.extractedPersons.isNotEmpty()) {
                put("extractedPersons", JSONArray(memory.extractedPersons))
            }
            if (memory.extractedTags.isNotEmpty()) {
                put("extractedTags", JSONArray(memory.extractedTags))
            }
            if (memory.personIds.isNotEmpty()) {
                put("personIds", JSONArray(memory.personIds))
            }
            if (memory.photos.isNotEmpty()) {
                put("photos", JSONArray().apply {
                    memory.photos.forEach { photo ->
                        put(JSONObject().apply {
                            put("id", photo.id)
                            put("url", photo.url)
                            photo.thumbnailUrl?.let { put("thumbnailUrl", it) }
                            photo.aiAnalysis?.let { put("aiAnalysis", it) }
                        })
                    }
                })
            }
        }
    }

    private fun personToJson(person: Person): JSONObject {
        return JSONObject().apply {
            put("id", person.id)
            put("name", person.name)
            put("relationship", person.relationship.name)
            person.nickname?.let { put("nickname", it) }
            person.phone?.let { put("phone", it) }
            person.email?.let { put("email", it) }
            person.memo?.let { put("memo", it) }
            person.profileImageUrl?.let { put("profileImageUrl", it) }
        }
    }

    private fun escapeCsv(value: String): String {
        val needsQuotes = value.contains(",") || value.contains("\"") || value.contains("\n")
        return if (needsQuotes) {
            "\"${value.replace("\"", "\"\"")}\""
        } else {
            value
        }
    }
}

data class ExportResult(
    val uri: Uri,
    val fileName: String,
    val mimeType: String,
    val itemCount: Int
)

enum class ExportFormat {
    JSON,
    CSV
}

File: ios/DailyMemory/Data/Service/ExportService.swift
import Foundation

/// Service for exporting memories to various formats
final class ExportService {
    static let shared = ExportService()

    private let fileManager = FileManager.default
    private let exportDirectory: URL

    private let dateFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    private let displayDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        return formatter
    }()

    private init() {
        let cacheDir = fileManager.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        exportDirectory = cacheDir.appendingPathComponent("exports", isDirectory: true)
        try? fileManager.createDirectory(at: exportDirectory, withIntermediateDirectories: true)
    }

    // MARK: - JSON Export

    /// Export memories to JSON format
    func exportToJson(memories: [Memory], persons: [Person] = []) async -> Result<ExportResult, Error> {
        await withCheckedContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async { [self] in
                do {
                    let timestamp = displayDateFormatter.string(from: Date())
                    let fileName = "dailymemory_export_\(timestamp).json"
                    let fileURL = exportDirectory.appendingPathComponent(fileName)

                    let exportData: [String: Any] = [
                        "exportDate": dateFormatter.string(from: Date()),
                        "version": "1.0",
                        "memoriesCount": memories.count,
                        "personsCount": persons.count,
                        "memories": memories.map { memoryToDict($0) },
                        "persons": persons.map { personToDict($0) }
                    ]

                    let jsonData = try JSONSerialization.data(withJSONObject: exportData, options: .prettyPrinted)
                    try jsonData.write(to: fileURL)

                    let result = ExportResult(
                        url: fileURL,
                        fileName: fileName,
                        mimeType: "application/json",
                        itemCount: memories.count
                    )

                    continuation.resume(returning: .success(result))
                } catch {
                    continuation.resume(returning: .failure(error))
                }
            }
        }
    }

    // MARK: - CSV Export

    /// Export memories to CSV format
    func exportToCsv(memories: [Memory]) async -> Result<ExportResult, Error> {
        await withCheckedContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async { [self] in
                do {
                    let timestamp = displayDateFormatter.string(from: Date())
                    let fileName = "dailymemory_export_\(timestamp).csv"
                    let fileURL = exportDirectory.appendingPathComponent(fileName)

                    var csvContent = "ID,Date,Time,Content,Category,Importance,Location,Amount,Persons,Tags,IsLocked\n"

                    let csvDateFormatter = DateFormatter()
                    csvDateFormatter.dateFormat = "yyyy-MM-dd"

                    let csvTimeFormatter = DateFormatter()
                    csvTimeFormatter.dateFormat = "HH:mm:ss"

                    for memory in memories {
                        let row = [
                            escapeCsv(memory.id),
                            csvDateFormatter.string(from: memory.recordedAt),
                            csvTimeFormatter.string(from: memory.recordedAt),
                            escapeCsv(memory.content),
                            memory.category.rawValue,
                            String(memory.importance),
                            escapeCsv(memory.extractedLocation ?? ""),
                            memory.extractedAmount.map { String($0) } ?? "",
                            escapeCsv(memory.extractedPersons.joined(separator: "; ")),
                            escapeCsv(memory.extractedTags.joined(separator: "; ")),
                            String(memory.isLocked)
                        ].joined(separator: ",")

                        csvContent += row + "\n"
                    }

                    try csvContent.write(to: fileURL, atomically: true, encoding: .utf8)

                    let result = ExportResult(
                        url: fileURL,
                        fileName: fileName,
                        mimeType: "text/csv",
                        itemCount: memories.count
                    )

                    continuation.resume(returning: .success(result))
                } catch {
                    continuation.resume(returning: .failure(error))
                }
            }
        }
    }

    // MARK: - Date Range Export

    /// Export memories for a specific date range
    func exportDateRange(
        memories: [Memory],
        startDate: Date,
        endDate: Date,
        format: ExportFormat
    ) async -> Result<ExportResult, Error> {
        let filteredMemories = memories.filter { memory in
            memory.recordedAt >= startDate && memory.recordedAt <= endDate
        }

        switch format {
        case .json:
            return await exportToJson(memories: filteredMemories)
        case .csv:
            return await exportToCsv(memories: filteredMemories)
        }
    }

    // MARK: - Cleanup

    /// Clean up old export files
    func cleanupExports() {
        guard let contents = try? fileManager.contentsOfDirectory(at: exportDirectory, includingPropertiesForKeys: [.contentModificationDateKey]) else {
            return
        }

        let oneDayAgo = Date().addingTimeInterval(-24 * 60 * 60)

        for url in contents {
            if let modificationDate = try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate,
               modificationDate < oneDayAgo {
                try? fileManager.removeItem(at: url)
            }
        }
    }

    // MARK: - Private Helpers

    private func memoryToDict(_ memory: Memory) -> [String: Any] {
        var dict: [String: Any] = [
            "id": memory.id,
            "content": memory.content,
            "category": memory.category.rawValue,
            "importance": memory.importance,
            "isLocked": memory.isLocked,
            "recordedAt": dateFormatter.string(from: memory.recordedAt),
            "createdAt": dateFormatter.string(from: memory.createdAt),
            "updatedAt": dateFormatter.string(from: memory.updatedAt)
        ]

        if let location = memory.extractedLocation {
            dict["extractedLocation"] = location
        }
        if let amount = memory.extractedAmount {
            dict["extractedAmount"] = amount
        }
        if let lat = memory.recordedLatitude {
            dict["latitude"] = lat
        }
        if let lng = memory.recordedLongitude {
            dict["longitude"] = lng
        }
        if !memory.extractedPersons.isEmpty {
            dict["extractedPersons"] = memory.extractedPersons
        }
        if !memory.extractedTags.isEmpty {
            dict["extractedTags"] = memory.extractedTags
        }
        if !memory.personIds.isEmpty {
            dict["personIds"] = memory.personIds
        }
        if !memory.photos.isEmpty {
            dict["photos"] = memory.photos.map { photo in
                var photoDict: [String: Any] = [
                    "id": photo.id,
                    "url": photo.url
                ]
                if let thumbnailUrl = photo.thumbnailUrl {
                    photoDict["thumbnailUrl"] = thumbnailUrl
                }
                if let aiAnalysis = photo.aiAnalysis {
                    photoDict["aiAnalysis"] = aiAnalysis
                }
                return photoDict
            }
        }

        return dict
    }

    private func personToDict(_ person: Person) -> [String: Any] {
        var dict: [String: Any] = [
            "id": person.id,
            "name": person.name,
            "relationship": person.relationship.rawValue
        ]

        if let nickname = person.nickname {
            dict["nickname"] = nickname
        }
        if let phone = person.phone {
            dict["phone"] = phone
        }
        if let email = person.email {
            dict["email"] = email
        }
        if let memo = person.memo {
            dict["memo"] = memo
        }
        if let profileImageUrl = person.profileImageUrl {
            dict["profileImageUrl"] = profileImageUrl
        }

        return dict
    }

    private func escapeCsv(_ value: String) -> String {
        let needsQuotes = value.contains(",") || value.contains("\"") || value.contains("\n")
        if needsQuotes {
            return "\"\(value.replacingOccurrences(of: "\"", with: "\"\""))\""
        }
        return value
    }
}

// MARK: - Models

struct ExportResult {
    let url: URL
    let fileName: String
    let mimeType: String
    let itemCount: Int
}

enum ExportFormat {
    case json
    case csv
}
